crates/subspace-networking/src/utils.rs

```diff
@@ -95,3 +97,150 @@ pub(crate) fn convert_multiaddresses(addresses: Vec<Multiaddr>) -> Vec<PeerAddre
+    async fn free_one(&self) {
+        let should_notify = {
+            let mut current = self.current.lock().await;
+            if current.usage > 0 {
+                current.usage -= 1;
+            } else {
+                panic!(
+                    "SemShared::free_one(): invalid free, current = {:?}",
+                    current
+                );
+            }
+
+            // Notify only if usage fell below the current capacity.
+            // For example: if the previous capacity was 100, and current capacity
+            // is 50, this will wait for usage to fall below 50 before any waiters
+            // are notified.
+            current.usage < current.capacity
+        };
+        if should_notify {
+            self.notify.notify_waiters();
+        }
+    }
+
+    // Expands the max capacity by delta, and notifies any waiters of the newly available
+    // free permits.
+    async fn expand(&self, delta: usize) {
+        {
+            let mut current = self.current.lock().await;
+            current.capacity += delta;
+        }
+        self.notify.notify_waiters();
+    }
+
+    // Shrinks the max capacity by delta
+    async fn shrink(&self, delta: usize) -> Result<(), String> {
+        let mut current = self.current.lock().await;
+        if current.capacity > delta {
+            current.capacity -= delta;
+            Ok(())
+        } else {
+            Err(format!(
+                "SemShared::::shrink(): invalid delta = {}, current = {:?}",
+                delta, current
+            ))
+        }
+    }
+}
+
+impl ResizableSemaphore {
+    pub(crate) fn new(capacity: usize) -> Self {
+        Self(Arc::new(SemShared::new(capacity)))
+    }
+
+    // Acquires a permit. Waits until a permit is available.
+    pub(crate) async fn acquire(&self) -> ResizableSemaphorePermit {
+        loop {
+            if self.0.alloc_one().await {
+                break;
+            }
+            self.0.notify.notified().await;
+        }
+        ResizableSemaphorePermit(self.0.clone())
+    }
+
+    // Acquires a permit, does not wait if no permits are available.
+    // Currently used only for testing.
+    #[cfg(test)]
+    pub(crate) async fn try_acquire(&self) -> Option<ResizableSemaphorePermit> {
+        if self.0.alloc_one().await {
+            Some(ResizableSemaphorePermit(self.0.clone()))
+        } else {
+            None
+        }
+    }
+
+    // Expands the capacity by the specified amount.
+    pub(crate) async fn expand(&self, delta: usize) {
+        self.0.expand(delta).await;
+    }
+
+    // Shrinks the capacity by the specified amount.
+    pub(crate) async fn shrink(&self, delta: usize) -> Result<(), String> {
+        self.0.shrink(delta).await
+    }
+}
+
+impl Drop for ResizableSemaphorePermit {
+    fn drop(&mut self) {
+        let state = self.0.clone();
+        tokio::spawn({
+            async move {
+                state.free_one().await;
+            }
+        });
```

Comment:
This is really unfortunate: spawning tasks for every permit. Looks like you can make it lock-free with atomics. `capacity` doesn't need to be under `Mutex`, it can be just `AtomicUsize` and updated with `Ordering::Relaxed` any time. For `usage` you can use a separate `AtomicUsize` with https://doc.rust-lang.org/stable/std/sync/atomic/struct.AtomicUsize.html#method.fetch_update for attempts to modify the value.

In fact, with atomics and correct memory ordering you should be able to fix race condition described above, though it is always tricky to deal with that stuff.

Reply:
Changed to sync Mutex